PlutoHelperAgent/PreferencesWindowController.m

```diff
@@ -0,0 +1,210 @@
+//
+//  PreferencesWindowController.m
+//  PlutoHelperAgent
+//
+//  Created by Dave Allison on 26/02/2018.
+//  Copyright © 2018 Guardian News & Media. All rights reserved.
+//
+
+#import "PreferencesWindowController.h"
+
+@interface PreferencesWindowController ()
+
+@end
+
+@implementation PreferencesWindowController
+
+
+- (void) grab_data_from_keychain {
+    
+    UInt32 pwLength = 0;
+    
+    void* pwData = NULL;
+    
+    SecKeychainItemRef itemRef = NULL;
+    
+    NSString* service = @"PlutoHelperAgent";
+    
+    OSStatus pwAccessStatus = SecKeychainFindGenericPassword(
+                                                     
+                                                     NULL,         // Search default keychains
+                                                     
+                                                     (UInt32)service.length,
+                                                     
+                                                     [service UTF8String],
+                                                     
+                                                     0,
+                                                     
+                                                     NULL,
+                                                     
+                                                     &pwLength,
+                                                     
+                                                     &pwData,
+                                                     
+                                                     &itemRef      // Get a reference this time
+                                                     
+                                                     );
+    
+    if (pwAccessStatus == errSecSuccess) {
+        
+        NSData* data = [NSData dataWithBytes:pwData length:pwLength];
+        
+        NSString* password = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+        
+        _PasswordText.stringValue = password;
+        
+        SecKeychainAttribute attrs2[] = {
+            
+            { kSecAccountItemAttr, 0, NULL }
+            
+        };
+        
+        SecKeychainAttributeList attributes2 = { 1, attrs2 };
+        
+        OSStatus unAccessStatus = SecKeychainItemCopyContent(itemRef, NULL, &attributes2, NULL, NULL);
+        
+        if (unAccessStatus == errSecSuccess) {
+            
+            NSLog(@"Username retrived from Apple Keychain");
+            
+            NSData* data8 = [NSData dataWithBytes:attributes2.attr->data length:attributes2.attr->length];
+            
+            NSString* usernamedata = [[NSString alloc] initWithData:data8 encoding:NSUTF8StringEncoding];
+            
+            _UsernameText.stringValue = usernamedata;
```

Comment:
I think that this is better written as [this setUsername@usernamedata]; - this way, you can use key-value observing from elsewhere in the app to get notified about updates to the member variables

Reply:
That does not work so leaving current syntax.